### ISSUE TITLE
Generate messages for ignored values in Ltac (solve, first); make rewrite non-existent lemma error fatal

### ISSUE THREAD

**Issue by jfehrle** [coq/coq#12478]
In Ltac, `solve` and `first` (and maybe other constructs) ignore non-tactic values that appear in the list of tactics to apply.  For example, `solve [ () | 1 | constructor ]` is equivalent to `solve [ constructor ]`.  This is likely a coding error, which would may not be so obvious if the non-tactic value comes from an Ltac variable.  We're not aware of a beneficial use for this behavior.

How about making this an error or warning?  If necessary, we could add a "strict" option (defaulting to on?) to control whether you get a message or the current silent treatment.

Separately, passing a non-existent lemma the `rewrite` tactic is not a fatal error, so an incorrect `rewrite` in a `first` or `solve` is silently ignored.  Seems like Ltac should be more helpful in this case.

@JasonGross can elaborate, if needed.  See https://github.com/coq/coq/pull/12103#discussion_r436323321

**Comment by jfehrle** [comment]
@JasonGross says:

> I believe the PR would consist of making "Tactic
expected" an exception that does not use user_err, and marking it as
critical / uncatchable by the tactic monad.

**Comment by Zimmi48** [comment]
I think the first problem can easily be fixed. OTOH, fixing second one (regarding `rewrite`) would be a clear change of semantics that seems out of scope for Ltac (because a "lemma" might also be a hypothesis from the context, which is looked for dynamically). I believe that Ltac2 already solves this issue.

**Comment by jfehrle** [comment]
These would both by dynamic checks.  For the first one the non-tactic value could come from a variable, not easy to check that statically.

For the second, clearly you'd check for both a lemma or a hypothesis.  That wouldn't be much of a change in semantics at all.  While it's great that Ltac2 may already solve the issue, it's premature to stop considering small improvements to Ltac1.  There's lots of existing Ltac1 code, there are no conversion tools or even a document describing how to convert it.

**Comment by Zimmi48** [comment]
> it's premature to stop considering small improvements to Ltac1

Agreed. I'm just not sure this qualifies as a small improvement. But nothing prevents from opening a PR to assess the compatibility breakage. If it is limited, why not?
